### ISSUE TITLE
Add Trivy SBOM scan for wave build containers

### ISSUE THREAD

**Issue by munishchouhan** [seqeralabs/wave#489]
from @ewels 
Seeing more and more people ask about SBOM documents for pipelines / containers (software bill of materials). It looks like Trivy can generate SBOMs. Is this something that we could get Wave to do automatically for every build? Then host it the same way that we do for the security scan results?
https://www.aquasec.com/blog/software-supply-chain-security-trivy-sbom/

**Comment by pditommaso** [comment]
Likely this can be implemented following the same "pattern" for the build logs: 

1. the command for the creation of the SBOM is included in the job carrying out the [Trivy scan](https://github.com/seqeralabs/wave/blob/master/src/main/groovy/io/seqera/wave/service/scan/ScanStrategy.groovy#L38-L45)
2. when the [scan complete](https://github.com/seqeralabs/wave/blob/master/src/main/groovy/io/seqera/wave/service/scan/ContainerScanServiceImpl.groovy#L87-L87) the SBOM file is collected and uploaded to a bucket, [similarly to a log](https://github.com/seqeralabs/wave/blob/master/src/main/groovy/io/seqera/wave/service/logs/BuildLogServiceImpl.groovy#L93-L93)
3. The endpoint `/builds/{id}/sbom` is added to retrieve the sbom content, similar to the [logs one](https://github.com/seqeralabs/wave/blob/master/src/main/groovy/io/seqera/wave/controller/BuildController.groovy#L64-L64)

**Comment by munishchouhan** [comment]
Working on this issue now, getting error with the sbom command, researching on that:
<img width="1906" alt="Screenshot 2024-06-27 at 19 48 04" src="https://github.com/seqeralabs/wave/assets/6575288/42dcd847-b639-4771-890f-a054ce51ea33">

**Comment by pditommaso** [comment]
Depends on https://github.com/seqeralabs/wave/issues/602

**Comment by pditommaso** [comment]
Does this require a separate job or it can be done along with the scan?

**Comment by munishchouhan** [comment]
> Does this require a separate job or it can be done along with the scan?
We can use existing scan job.
we need to add `--list-all-pkgs` flag in our existing scan command which will also add sbom in the output json

**Comment by munishchouhan** [comment]
like this
`docker run --rm aquasec/trivy:0.55.0  -q image --format json --list-all-pkgs ubuntu:20.04 `
it will have new package section which will contain
```
"Results": [
    {
      "Target": "ubuntu:20.04 (ubuntu 20.04)",
      "Class": "os-pkgs",
      "Type": "ubuntu",
      "Packages": [
        {
          "ID": "adduser@3.118ubuntu2",
          "Name": "adduser",
          "Identifier": {
            "PURL": "pkg:deb/ubuntu/adduser@3.118ubuntu2?arch=all\u0026distro=ubuntu-20.04",
            "UID": "5122302fc4653274"
          },
          "Version": "3.118ubuntu2",
          "Arch": "all",
          "SrcName": "adduser",
          "SrcVersion": "3.118ubuntu2",
          "Licenses": [
            "GPL-2.0"
          ],
          "Maintainer": "Ubuntu Core Developers \u003cubuntu-devel-discuss@lists.ubuntu.com\u003e",
          "DependsOn": [
            "debconf@1.5.73",
            "passwd@1:4.8.1-1ubuntu5.20.04.5"
          ], 
```

**Comment by pditommaso** [comment]
Nice. @ewels is what you were expecting by this?

**Comment by munishchouhan** [comment]
There are two main industry standards for SBOMs:

1. Software Package Data Exchange (SPDX and SPDX Lite), an ISO standard hosted by the Linux Foundation, which outlines the components, licenses, and copyrights associated with a software package.
2. CycloneDX, an open source, lightweight SBOM standard, which is used in application security and supply chain analysis and originated from the Open Web Application Security Project (OWASP).

**Comment by pditommaso** [comment]
Can both of them be created with the same scan command?

**Comment by munishchouhan** [comment]
> Can both of them be created with the same scan command?

no we need provide the specific format for it
in case of CycloneDX, we get vulnerabilities too, but not in case of SPDX

if we want SPDX, then we need two jobs

there is PR which add support for security scan in SPDX, but it don't see it in the results
https://github.com/aquasecurity/trivy/pull/7213

examples:

1.  CycloneDX
` docker run --rm aquasec/trivy:0.55.0  -q image --format cyclonedx --list-all-pkgs --scanners vuln ubuntu:20.04 `

2. spdx 
`docker run --rm aquasec/trivy:0.55.0  -q image --format spdx --list-all-pkgs --scanners vuln ubuntu:20.04`

`docker run --rm aquasec/trivy:0.55.0  -q image --format spdx-json --list-all-pkgs --scanners vuln ubuntu:20.04`

**Comment by munishchouhan** [comment]
correction: SPDX also contains vulnerabilities, but format is different:


```
"externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:deb/ubuntu/libc-bin@2.39-0ubuntu8.2?arch=amd64\u0026distro=ubuntu-24.04"
        },
        {
          "referenceCategory": "SECURITY",
          "referenceType": "advisory",
          "referenceLocator": "https://avd.aquasec.com/nvd/cve-2016-20013"
        }
      ],
```

**Comment by ewels** [comment]
Any idea what other services provide? eg. Docker / Quay etc? Could do worse than comparing ourselves against them.

**Comment by pditommaso** [comment]
Hi, thanks for your interest in the project and your feedback. As a user how would you expect to consume the SBOM report? just accessing and downloading the JSON file?

**Comment by thefabnab** [comment]
Yup exactly. Since the SOBM a build artifact (at least that's what it seems like to me) exposing some kind of endpoint to serve these as a json sounds great to me.
